cmd/packer-sdc/internal/fix/cmd.go

```diff
@@ -0,0 +1,157 @@
+
+package fix
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/diff"
+)
+
+const cmdPrefix string = "fix"
+
+// Command is the base entry for the fix sub-command.
+type Command struct {
+	Dir  string
+	Diff bool
+}
+
+// Flags contain the default flags for the fix sub-command.
+func (cmd *Command) Flags() *flag.FlagSet {
+	fs := flag.NewFlagSet(cmdPrefix, flag.ExitOnError)
+	fs.BoolVar(&cmd.Diff, "diff", false, "if set prints the differences a rewrite would introduce.")
+	return fs
+}
+
+// Help displays usage for the command.
+func (cmd *Command) Help() string {
+	var s strings.Builder
+	for _, fix := range availableFixes {
+		s.WriteString(fmt.Sprintf("  %s\t\t%s\n", fix.name, fix.description))
+	}
+
+	helpText := `
+Usage: packer-sdc fix [options] directory
+
+  Fix rewrites parts of the plugin codebase to address known issues or
+  common workarounds used within plugins consuming the Packer plugin SDK.
+
+Options:
+  -diff		If the -diff flag is set fix prints the differences an applied fix would introduce.
+
+Available fixes:
+%s`
+	return fmt.Sprintf(helpText, s.String())
+}
+
+// Run executes the command
+func (cmd *Command) Run(args []string) int {
+	if err := cmd.run(args); err != nil {
+		log.Printf("%v", err)
+		return 1
+	}
+	return 0
+}
+
+func (cmd *Command) run(args []string) error {
+	f := cmd.Flags()
+	err := f.Parse(args)
+	if err != nil {
+		return errors.New("unable to parse flags for fix command")
+	}
+
+	if f.NArg() != 1 {
+		fmt.Println(cmd.Help())
+		return nil
+	}
+
+	dir := f.Arg(0)
+	if dir == "." || dir == "./..." {
+		dir, _ = os.Getwd()
+	}
+
+	info, err := os.Stat(dir)
+	if err != nil && os.IsNotExist(err) {
+		return errors.New("a plugin root directory must be specified or a dot for the current directory")
+	}
+
+	if !info.IsDir() {
+		return errors.New("a plugin root directory must be specified or a dot for the current directory")
+	}
+
+	dir, err = filepath.Abs(dir)
+	if err != nil {
+		return errors.New("unable to determine the absolute path for the provided plugin root directory")
+	}
+	cmd.Dir = dir
+
+	return processFiles(cmd.Dir, cmd.Diff)
+}
+
+func (cmd *Command) Synopsis() string {
+	return "Rewrites parts of the plugin codebase to address known issues or common workarounds within plugins consuming the Packer plugin SDK."
+}
+
+func processFiles(rootDir string, showDiff bool) error {
+	srcFiles := make(map[string][]byte)
+	fixedFiles := make(map[string][]byte)
+
+	var hasErrors error
+	for _, f := range availableFixes {
+		matches, err := f.scan(rootDir)
+		if err != nil {
+			return fmt.Errorf("failed to apply %s fix: %s", f.name, err)
+		}
+
+		//matches contains all files to apply the said fix on
+		for _, filename := range matches {
+			if _, ok := srcFiles[filename]; !ok {
+				// TODO handle error
+				bs, _ := os.ReadFile(filename)
+				srcFiles[filename] = bytes.Clone(bs)
+			}
+
+			fixedData, ok := fixedFiles[filename]
```

Comment:
Why you fix the data again?

Reply:
So it is not really an issue since we have one fix but if there are multiple fixes that would get applied to the same file  we need to make sure that the fixes are accumulative and not modifying the same unfixed src file. 

To do ☝️ I store the already fixed data into the fixedFiles map and before applying another fix I check if the contents of the file have changed if so I grab the fixed file before applying the next fix. If we've never seen the file before or it has not been fixed previously we copy the original source into fixedData and only store within the fixedFiles map if a fix has been applied. 

Please let me know if I explained that well or if you're still unclear on why

Reply:
oh yeah, I see now! First time I looked I didn't realized that fixedFiles is used for all fixes. Makes sense to me, thanks for explaining!